### PR TITLE
Add vLLM backend

### PR DESCRIPTION
**Pull request by leseb**
See individual commits.

**Checklist:**

- [x] **Commit Message Formatting**: Commit titles and messages follow guidelines in the
  [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary).
- [x] [Changelog](https://github.com/instructlab/instructlab/blob/main/CHANGELOG.md) updated with breaking and/or notable changes for the next minor release.
- [ ] Documentation has been updated, if necessary.
- [x] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.

### CODE REVIEW

**Comment by leseb** [comment]
This contains everything related to the vLLM integration work.

**Comment by tiran** [comment]
vLLM does not support Python 3.12 and therefore does not have binaries for 3.12 on PyPI. It also has a really strange build system. vLLM build systems needs requires a PyTorch installation to build vLLM for a platform.

Basically vLLM only works out of the box on Linux x86_64 with CUDA on Python >=3.9 <= 3.11. Everything else is not working at the moment or cannot be expressed as a simple requirement in wheels.

**Comment by tiran** [comment]
And you also need to install a bunch of packages manually for each GPU arch, e.g. https://github.com/vllm-project/vllm/blob/main/requirements-cuda.txt for CUDA and https://github.com/vllm-project/vllm/blob/main/requirements-rocm.txt for ROCm.

Except do NOT install Ray. It's only needed for multi-node inference and is an even greater PITB to install correctly.

**Comment by n1hility** [comment]
> And you also need to install a bunch of packages manually for each GPU arch, e.g. https://github.com/vllm-project/vllm/blob/main/requirements-cuda.txt for CUDA and https://github.com/vllm-project/vllm/blob/main/requirements-rocm.txt for ROCm.
> 
> Except do NOT install Ray. It's only needed for multi-node inference and is an even greater PITB to install correctly.

Previously we talked about making vllm an optional dep which people could pip install instructlab[vllm].  I think that probably makes sense still. wdyt @tiran ?

File: src/instructlab/model/backends/backends.py
Comment (alimaredia) [inline src/instructlab/model/backends/backends.py:141]: As a follow up PR, we could auto detect `.safetensors` files, and the presence of certain specific `.json` files existing. I think a follow up issue for this could get created as soon as the PR gets merged, and we don't hold up on it.

Comment (leseb) [reply src/instructlab/model/backends/backends.py:141]: Sounds good as a follow-up item to me. Thanks!

File: scripts/basic-workflow-tests.sh
Comment (nathan-weinberg) [inline scripts/basic-workflow-tests.sh:308]: Good thinking adding this!

File: src/instructlab/data/generate.py
Comment (nathan-weinberg) [inline src/instructlab/data/generate.py:175]:
```suggestion
        model_path = pathlib.Path(ctx.obj.config.serve.model_path)
```

File: src/instructlab/data/generate.py
Comment (nathan-weinberg) [inline src/instructlab/data/generate.py:187]: nit
```suggestion
            # Initialize the llama server
```

File: src/instructlab/data/generate.py
Comment (nathan-weinberg) [inline src/instructlab/data/generate.py:194]:
```suggestion
                num_threads=None,  # exists only as a flag, not a config value
```

File: src/instructlab/data/generate.py
Comment (nathan-weinberg) [inline src/instructlab/data/generate.py:201]: nit
```suggestion
            # Initialize the vllm server
```

**Review by nathan-weinberg — COMMENTED** [review]
Will do another pass later but this is looking good @leseb @alimaredia!

File: requirements-vllm-cuda.txt
Comment (tiran) [inline requirements-vllm-cuda.txt:5]: You also have to include the dependencies from https://github.com/vllm-project/vllm/blob/main/requirements-cuda.txt (except for Ray). At least `vllm-flash-attn` should be installed to improve performance.

Comment (leseb) [reply requirements-vllm-cuda.txt:5]: Yes thanks, I was just testing out something on the runner first, I'll include these.

File: src/instructlab/model/backends/backends.py
Comment (leseb) [reply src/instructlab/model/backends/backends.py:141]: https://github.com/instructlab/instructlab/issues/1457

File: src/instructlab/data/generate.py
Comment (leseb) [reply src/instructlab/data/generate.py:196]: It's effectively "instantiating" `backend_instance` as an object.

File: src/instructlab/data/generate.py
Comment (leseb) [reply src/instructlab/data/generate.py:210]: Ditto

**Review by nathan-weinberg — DISMISSED** [review]

File: CHANGELOG.md
Comment (alimaredia) [inline CHANGELOG.md:8]: @leseb This should include `vllm` as a supported backend, right?

File: src/instructlab/model/backends/vllm.py
Comment (n1hility) [inline src/instructlab/model/backends/vllm.py:67]: dead code?

File: CHANGELOG.md
Comment (leseb) [reply CHANGELOG.md:8]: updating

File: src/instructlab/model/backends/vllm.py
Comment (leseb) [reply src/instructlab/model/backends/vllm.py:67]: yes :(

**Review by n1hility — APPROVED** [review]

**Review by nathan-weinberg — APPROVED** [review]

File: src/instructlab/model/backends/vllm.py
Comment (alimaredia) [inline src/instructlab/model/backends/vllm.py:145]: @leseb are `eos_token` and `bos_token` just here for the TODO? They don't seem to lead anywhere

Comment (leseb) [reply src/instructlab/model/backends/vllm.py:145]: Yes, just TODO.

**Review by alimaredia — APPROVED** [review]

**Comment by cdoern** [comment]
this probably should have come after #1370

**Pull request merged**
